File: cpp/src/Freeze/EvictorI.cpp
// **********************************************************************
//
// Copyright (c) 2003-2017 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

#include <Freeze/EvictorI.h>
#include <Freeze/Initialize.h>
#include <Freeze/Util.h>
#include <Freeze/EvictorIteratorI.h>
#include <Freeze/PingObject.h>

#include <IceUtil/IceUtil.h>

#include <IceUtil/StringConverter.h>

#include <typeinfo>

using namespace std;
using namespace Freeze;
using namespace Ice;
using namespace IceUtil;

//
// Static members
//

string Freeze::EvictorIBase::defaultDb = "$default";
string Freeze::EvictorIBase::indexPrefix = "$index:";


//
// DeactivateController
//

Freeze::DeactivateController::Guard::Guard(const DeactivateController& controller) :
    _controller(const_cast<DeactivateController&>(controller))
{
    Lock sync(controller);
    if(controller._deactivated || _controller._deactivating)
    {
        throw EvictorDeactivatedException(__FILE__, __LINE__);
    }
    _controller._guardCount++;
}

Freeze::DeactivateController::Guard::~Guard()
{
    Lock sync(_controller);
    _controller._guardCount--;
    if(_controller._deactivating && _controller._guardCount == 0)
    {
        //
        // Notify all the threads -- although we only want to
        // reach the thread doing the deactivation.
        //
        _controller.notifyAll();
    }
}

Freeze::DeactivateController::DeactivateController(EvictorIBase* evictor) :
    _evictor(evictor),
    _deactivating(false),
    _deactivated(false),
    _guardCount(0)
{
}

bool
Freeze::DeactivateController::deactivated() const
{
    Lock sync(*this);
    return _deactivated || _deactivating;
}

bool
Freeze::DeactivateController::deactivate()
{
    Lock sync(*this);

    if(_deactivated)
    {
        return false;
    }

    if(_deactivating)
    {
        //
        // Wait for deactivated
        //
        while(!_deactivated)
        {
            wait();
        }
        return false;
    }
    else
    {
        _deactivating = true;
        while(_guardCount > 0)
        {
            if(_evictor->trace() >= 1)
            {
                Trace out(_evictor->communicator()->getLogger(), "Freeze.Evictor");
                out << "Waiting for " << _guardCount << " threads to complete before starting deactivation.";
            }

            wait();
        }

        if(_evictor->trace() >= 1)
        {
            Trace out(_evictor->communicator()->getLogger(), "Freeze.Evictor");
            out << "Starting deactivation.";
        }
        return true;
    }
}

void
Freeze::DeactivateController::deactivationComplete()
{
    if(_evictor->trace() >= 1)
    {
        Trace out(_evictor->communicator()->getLogger(), "Freeze.Evictor");
        out << "Deactivation complete.";
    }

    Lock sync(*this);
    _deactivated = true;
    _deactivating = false;
    notifyAll();
}


//
// EvictorIBase
//

Freeze::EvictorIBase::EvictorIBase(const ObjectAdapterPtr& adapter,
                                   const string& envName,
                                   DbEnv* dbEnv,
                                   const string& filename,
                                   const FacetTypeMap& facetTypes,
                                   const ServantInitializerPtr& initializer,
                                   bool createDb) :
    _evictorSize(10),
    _facetTypes(facetTypes),
    _deactivateController(this),
    _adapter(adapter),
    _communicator(adapter->getCommunicator()),
    _initializer(initializer),
    _dbEnv(SharedDbEnv::get(_communicator, envName, dbEnv)),
    _filename(filename),
    _createDb(createDb),
    _trace(0),
    _txTrace(0),
    _pingObject(new PingObject)
{
    _encoding = _dbEnv->getEncoding();

    _trace = _communicator->getProperties()->getPropertyAsInt("Freeze.Trace.Evictor");
    _txTrace = _communicator->getProperties()->getPropertyAsInt("Freeze.Trace.Transaction");
    _deadlockWarning = (_communicator->getProperties()->getPropertyAsInt("Freeze.Warn.Deadlocks") > 0);
}


void
Freeze::EvictorIBase::setSize(Int evictorSize)
{
    DeactivateController::Guard deactivateGuard(_deactivateController);

    Lock sync(*this);

    //
    // Ignore requests to set the evictor size to values smaller than zero.
    //
    if(evictorSize < 0)
    {
        return;
    }

    //
    // Update the evictor size.
    //
    _evictorSize = static_cast<size_t>(evictorSize);

    //
    // Evict as many elements as necessary.
    //
    evict();
}

Int
Freeze::EvictorIBase::getSize()
{
    Lock sync(*this);
    return static_cast<Int>(_evictorSize);
}


Ice::ObjectPrx
Freeze::EvictorIBase::add(const ObjectPtr& servant, const Identity& ident)
{
    return addFacet(servant, ident, "");
}

Ice::ObjectPtr
Freeze::EvictorIBase::remove(const Identity& ident)
{
    return removeFacet(ident, "");
}

bool
Freeze::EvictorIBase::hasObject(const Identity& ident)
{
    return hasFacet(ident, "");
}

ObjectPtr
Freeze::EvictorIBase::locate(const Current& current, LocalObjectPtr& cookie)
{
    //
    // Special ice_ping() handling
    //
    if(current.operation == "ice_ping")
    {
        if(hasFacet(current.id, current.facet))
        {
            if(_trace >= 3)
            {
                Trace out(_communicator->getLogger(), "Freeze.Evictor");
                out << "ice_ping found \"" << _communicator->identityToString(current.id)
                    << "\" with facet \"" << current.facet + "\"";
            }

            cookie = 0;
            return _pingObject;
        }
        else if(hasAnotherFacet(current.id, current.facet))
        {
            if(_trace >= 3)
            {
                Trace out(_communicator->getLogger(), "Freeze.Evictor");
                out << "ice_ping raises FacetNotExistException for \"" << _communicator->identityToString(current.id)
                    << "\" with facet \"" << current.facet + "\"";
            }
            throw FacetNotExistException(__FILE__, __LINE__);
        }
        else
        {
            if(_trace >= 3)
            {
                Trace out(_communicator->getLogger(), "Freeze.Evictor");
                out << "ice_ping will raise ObjectNotExistException for \""
                    << _communicator->identityToString(current.id)  << "\" with facet \"" << current.facet + "\"";
            }
            return 0;
        }
    }

    ObjectPtr result = locateImpl(current, cookie);

    if(result == 0)
    {
        if(hasAnotherFacet(current.id, current.facet))
        {
            throw FacetNotExistException(__FILE__, __LINE__);
        }
    }
    return result;
}


void
Freeze::EvictorIBase::initialize(const Identity& ident, const string& facet, const ObjectPtr& servant)
{
    if(_initializer != 0)
    {
        _initializer->initialize(_adapter, ident, facet, servant);
    }
}

void
Freeze::EvictorIBase::updateStats(Statistics& stats, IceUtil::Int64 time)
{
    IceUtil::Int64 diff = time - (stats.creationTime + stats.lastSaveTime);
    if(stats.lastSaveTime == 0)
    {
        stats.lastSaveTime = diff;
        stats.avgSaveTime = diff;
    }
    else
    {
        stats.lastSaveTime = time - stats.creationTime;
        stats.avgSaveTime = static_cast<IceUtil::Int64>(stats.avgSaveTime * 0.95 + diff * 0.05);
    }
}


const string&
Freeze::EvictorIBase::filename() const
{
    return _filename;
}

vector<string>
Freeze::EvictorIBase::allDbs() const
{
    vector<string> result;

    try
    {
        Db db(_dbEnv->getEnv(), 0);

        //
        // Berkeley DB expects file paths to be UTF8 encoded.
        //
        db.open(0, nativeToUTF8(_filename, IceUtil::getProcessStringConverter()).c_str(), 0, DB_UNKNOWN,
                DB_RDONLY, 0);

        Dbc* dbc = 0;
        db.cursor(0, &dbc, 0);

        Dbt dbKey;
        dbKey.set_flags(DB_DBT_MALLOC);

        Dbt dbValue;
        dbValue.set_flags(DB_DBT_USERMEM | DB_DBT_PARTIAL);

        bool more = true;
        while(more)
        {
            more = (dbc->get(&dbKey, &dbValue, DB_NEXT) == 0);
            if(more)
            {
                string dbName(static_cast<char*>(dbKey.get_data()), dbKey.get_size());

                if(dbName.find(indexPrefix) != 0)
                {
                    result.push_back(dbName);
                }
                free(dbKey.get_data());
            }
        }

        dbc->close();
        db.close(0);
    }
    catch(const DbException& dx)
    {
        if(dx.get_errno() != ENOENT)
        {
            DatabaseException ex(__FILE__, __LINE__);
            ex.message = dx.what();
            throw ex;
        }
    }

    return result;
}

File: java/src/Freeze/src/main/java/Freeze/EvictorI.java
// **********************************************************************
//
// Copyright (c) 2003-2018 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

package Freeze;

abstract class EvictorI implements Evictor
{
    //
    // The deactivate controller is used by the implementation of all public
    // operations to ensure that deactivate() (which closes/clears various
    // Berkeley DB objects) is not called during these operations.
    // Note that the only threads that may perform such concurrent calls
    // are threads other than the dispatch threads of the associated adapter.
    //
    class DeactivateController
    {
        synchronized void
        activate()
        {
            assert !_activated;
            _activated = true;
        }

        synchronized void
        lock()
        {
            assert _activated;

            if(_deactivated || _deactivating)
            {
                throw new EvictorDeactivatedException();
            }
            _guardCount++;
        }

        synchronized void
        unlock()
        {
            assert _activated;

            _guardCount--;
            if(_deactivating && _guardCount == 0)
            {
                //
                // Notify all the threads -- although we only want to
                // reach the thread doing the deactivation.
                //
                notifyAll();
            }
        }

        synchronized boolean
        deactivated()
        {
            return !_activated || _deactivated;
        }

        synchronized boolean
        deactivate()
        {
            assert _activated;

            if(_deactivated)
            {
                return false;
            }

            if(_deactivating)
            {
                //
                // Wait for deactivated
                //
                while(!_deactivated)
                {
                    try
                    {
                        wait();
                    }
                    catch(InterruptedException e)
                    {
                        // Ignored
                    }
                }
                return false;
            }
            else
            {
                _deactivating = true;
                while(_guardCount > 0)
                {
                    if(_trace >= 1)
                    {
                        _communicator.getLogger().trace("Freeze.Evictor",  "Waiting for " + _guardCount +
                            " threads to complete before starting deactivation.");
                    }

                    try
                    {
                        wait();
                    }
                    catch(InterruptedException e)
                    {
                        // Ignored
                    }
                }

                if(_trace >= 1)
                {
                    _communicator.getLogger().trace("Freeze.Evictor", "Starting deactivation.");
                }
                return true;
            }
        }

        synchronized void
        deactivationComplete()
        {
            if(_trace >= 1)
            {
                _communicator.getLogger().trace("Freeze.Evictor", "Deactivation complete.");
            }

            _deactivated = true;
            _deactivating = false;
            notifyAll();
        }

        private boolean _activated = false;
        private boolean _deactivating = false;
        private boolean _deactivated = false;
        private int _guardCount = 0;
    }

    static final String defaultDb = "$default";
    static final String indexPrefix = "$index:";

    @Override
    public Ice.ObjectPrx
    add(Ice.Object servant, Ice.Identity ident)
    {
        return addFacet(servant, ident, "");
    }

    @Override
    public Ice.Object
    remove(Ice.Identity ident)
    {
        return removeFacet(ident, "");
    }

    @Override
    public boolean
    hasObject(Ice.Identity ident)
    {
        return hasFacet(ident, "");
    }

    @Override
    public Ice.Object
    locate(Ice.Current current, Ice.LocalObjectHolder cookie)
    {
        //
        // Special ice_ping() handling
        //
        if(current.operation != null && current.operation.equals("ice_ping"))
        {
            if(hasFacet(current.id, current.facet))
            {
                if(_trace >= 3)
                {
                    _communicator.getLogger().trace(
                        "Freeze.Evictor", "ice_ping found \"" + _communicator.identityToString(current.id) +
                        "\" with facet \"" + current.facet + "\"");
                }

                cookie.value = null;
                return _pingObject;
            }
            else if(hasAnotherFacet(current.id, current.facet))
            {
                if(_trace >= 3)
                {
                    _communicator.getLogger().trace(
                        "Freeze.Evictor", "ice_ping raises FacetNotExistException for \"" +
                        _communicator.identityToString(current.id)  + "\" with facet \"" + current.facet + "\"");
                }

                throw new Ice.FacetNotExistException();
            }
            else
            {
                if(_trace >= 3)
                {
                    _communicator.getLogger().trace(
                        "Freeze.Evictor", "ice_ping will raise ObjectNotExistException for \"" +
                        _communicator.identityToString(current.id)  + "\" with facet \"" + current.facet + "\"");
                }

                return null;
            }
        }

        Ice.Object result = locateImpl(current, cookie);

        if(result == null)
        {
            if(hasAnotherFacet(current.id, current.facet))
            {
                throw new Ice.FacetNotExistException(current.id, current.facet, current.operation);
            }
        }
        return result;
    }

    @Override
    synchronized public void
    setSize(int evictorSize)
    {
        _deactivateController.lock();
        try
        {
            //
            // Ignore requests to set the evictor size to values smaller than zero.
            //
            if(evictorSize < 0)
            {
                return;
            }

            //
            // Update the evictor size.
            //
            _evictorSize = evictorSize;

            //
            // Evict as many elements as necessary.
            //
            evict();
        }
        finally
        {
            _deactivateController.unlock();
        }
    }

    @Override
    synchronized public int
    getSize()
    {
        return _evictorSize;
    }

    @Override
    public EvictorIterator
    getIterator(String facet, int batchSize)
    {
        _deactivateController.lock();
        try
        {
            if(facet == null)
            {
                facet = "";
            }
            TransactionI tx = beforeQuery();
            return new EvictorIteratorI(findStore(facet, false), tx, batchSize);
        }
        finally
        {
            _deactivateController.unlock();
        }
    }

    abstract protected boolean hasAnotherFacet(Ice.Identity ident, String facet);

    abstract protected Object createEvictorElement(Ice.Identity ident, ObjectRecord rec, ObjectStore store);

    abstract protected Ice.Object locateImpl(Ice.Current current, Ice.LocalObjectHolder cookie);

    abstract protected void evict();

    protected void
    closeDbEnv()
    {
        assert _dbEnv != null;
        for(ObjectStore store : _storeMap.values())
        {
            store.close();
        }
        _dbEnv.close();
        _dbEnv = null;
    }

    protected synchronized ObjectStore
    findStore(String facet, boolean createIt)
    {
        ObjectStore os = _storeMap.get(facet);

        if(os == null && createIt)
        {
            String facetType = _facetTypes.get(facet);
            os = new ObjectStore(facet, facetType, true, this, new java.util.LinkedList<Index>(), false);
            _storeMap.put(facet, os);
        }
        return os;
    }

    protected void
    initialize(Ice.Identity ident, String facet, Ice.Object servant)
    {
        if(_initializer != null)
        {
            _initializer.initialize(_adapter, ident, facet, servant);
        }
    }

    protected
    EvictorI(Ice.ObjectAdapter adapter, String envName, com.sleepycat.db.Environment dbEnv, String filename,
             java.util.Map<String, String> facetTypes, ServantInitializer initializer, Index[] indices,
             boolean createDb)
    {
        _adapter = adapter;
        _communicator = adapter.getCommunicator();
        _initializer = initializer;
        _filename = filename;
        _createDb = createDb;
        _facetTypes = facetTypes == null ? new java.util.HashMap<String, String>() :
            new java.util.HashMap<String, String>(facetTypes);

        _dbEnv = SharedDbEnv.get(_communicator, envName, dbEnv);
        _encoding = _dbEnv.getEncoding();

        _trace = _communicator.getProperties().getPropertyAsInt("Freeze.Trace.Evictor");
        _txTrace = _communicator.getProperties().getPropertyAsInt("Freeze.Trace.Transaction");
        _deadlockWarning = _communicator.getProperties().getPropertyAsInt("Freeze.Warn.Deadlocks") > 0;

        _errorPrefix = "Freeze Evictor DbEnv(\"" + envName + "\") Db(\"" + _filename + "\"): ";

        String propertyPrefix = "Freeze.Evictor." + envName + '.' + _filename;

        boolean populateEmptyIndices =
            _communicator.getProperties().getPropertyAsIntWithDefault(propertyPrefix + ".PopulateEmptyIndices", 0) > 0;

        //
        // Instantiate all Dbs in 2 steps:
        // (1) iterate over the indices and create ObjectStore with indices
        // (2) open ObjectStores without indices
        //

        java.util.List<String> dbs = allDbs();
        //
        // Add default db in case it's not there
        //
        dbs.add(defaultDb);

        if(indices != null)
        {
            for(int i = 0; i < indices.length; ++i)
            {
                String facet = indices[i].facet();

                if(_storeMap.get(facet) == null)
                {
                    java.util.List<Index> storeIndices = new java.util.LinkedList<Index>();
                    for(int j = i; j < indices.length; ++j)
                    {
                        if(indices[j].facet().equals(facet))
                        {
                            storeIndices.add(indices[j]);
                        }
                    }

                    String facetType = _facetTypes.get(facet);
                    ObjectStore store = new ObjectStore(facet, facetType,_createDb, this, storeIndices,
                                                        populateEmptyIndices);
                    _storeMap.put(facet, store);
                }
            }
        }

        for(String facet : dbs)
        {
            if(facet.equals(defaultDb))
            {
                facet = "";
            }

            if(_storeMap.get(facet) == null)
            {
                String facetType = _facetTypes.get(facet);

                ObjectStore store = new ObjectStore(facet, facetType, _createDb, this,
                                                    new java.util.LinkedList<Index>(), populateEmptyIndices);

                _storeMap.put(facet, store);
            }
        }
        _deactivateController.activate();
    }

    protected
    EvictorI(Ice.ObjectAdapter adapter, String envName, String filename, java.util.Map<String, String> facetTypes,
             ServantInitializer initializer, Index[] indices, boolean createDb)
    {
        this(adapter, envName, null, filename, facetTypes, initializer, indices, createDb);
    }

    abstract TransactionI beforeQuery();

    static void
    updateStats(Statistics stats, long time)
    {
        long diff = time - (stats.creationTime + stats.lastSaveTime);
        if(stats.lastSaveTime == 0)
        {
            stats.lastSaveTime = diff;
            stats.avgSaveTime = diff;
        }
        else
        {
            stats.lastSaveTime = time - stats.creationTime;
            stats.avgSaveTime = (long)(stats.avgSaveTime * 0.95 + diff * 0.05);
        }
    }

    final DeactivateController
    deactivateController()
    {
        return _deactivateController;
    }

    final Ice.Communicator
    communicator()
    {
        return _communicator;
    }

    final Ice.EncodingVersion
    encoding()
    {
        return _encoding;
    }

    final SharedDbEnv
    dbEnv()
    {
        return _dbEnv;
    }

    final String
    filename()
    {
        return _filename;
    }

    final String
    errorPrefix()
    {
        return _errorPrefix;
    }

    final boolean
    deadlockWarning()
    {
        return _deadlockWarning;
    }

    final int
    trace()
    {
        return _trace;
    }

    private java.util.List<String>
    allDbs()
    {
        java.util.List<String> result = new java.util.LinkedList<String>();

        com.sleepycat.db.Database db = null;
        com.sleepycat.db.Cursor dbc = null;

        try
        {
            com.sleepycat.db.DatabaseConfig config = new com.sleepycat.db.DatabaseConfig();
            config.setType(com.sleepycat.db.DatabaseType.UNKNOWN);
            config.setReadOnly(true);
            db = _dbEnv.getEnv().openDatabase(null, _filename, null, config);

            dbc = db.openCursor(null, null);

            com.sleepycat.db.DatabaseEntry key = new com.sleepycat.db.DatabaseEntry();
            com.sleepycat.db.DatabaseEntry value = new com.sleepycat.db.DatabaseEntry();

            boolean more = true;
            while(more)
            {
                more = (dbc.getNext(key, value, null) == com.sleepycat.db.OperationStatus.SUCCESS);
                if(more)
                {
                    //
                    // Assumes Berkeley-DB encodes the db names in UTF-8!
                    //
                    String dbName = new String(key.getData(), 0, key.getSize(), "UTF8");

                    if(!dbName.startsWith(indexPrefix))
                    {
                        result.add(dbName);
                    }
                }
            }

            dbc.close();
            dbc = null;
            db.close();
            db = null;
        }
        catch(java.io.UnsupportedEncodingException ix)
        {
            throw new DatabaseException(_errorPrefix + "cannot decode database names", ix);
        }
        catch(java.io.FileNotFoundException ix)
        {
            //
            // New file
            //
        }
        catch(com.sleepycat.db.DatabaseException dx)
        {
            throw new DatabaseException(_errorPrefix + "Db.open: " + dx.getMessage(), dx);
        }
        finally
        {
            if(dbc != null)
            {
                try
                {
                    dbc.close();
                }
                catch(com.sleepycat.db.DatabaseException dx)
                {
                    // Ignored
                }
            }

            if(db != null)
            {
                try
                {
                    db.close();
                }
                catch(com.sleepycat.db.DatabaseException dx)
                {
                    // Ignored
                }
            }
        }

        return result;
    }

    static void
    checkIdentity(Ice.Identity ident)
    {
        if(ident.name == null || ident.name.length() == 0)
        {
            throw new Ice.IllegalIdentityException(ident);
        }
    }

    static void
    checkServant(Ice.Object servant)
    {
        if(servant == null)
        {
            throw new Ice.IllegalServantException("cannot add null servant to Freeze Evictor");
        }
    }

    protected int _evictorSize = 10;

    protected final java.util.Map<String, ObjectStore> _storeMap = new java.util.HashMap<String, ObjectStore>();
    private final java.util.Map<String, String> _facetTypes;

    protected final Ice.ObjectAdapter _adapter;
    protected final Ice.Communicator _communicator;
    protected final Ice.EncodingVersion _encoding;

    protected final ServantInitializer _initializer;

    protected SharedDbEnv  _dbEnv;

    protected final String _filename;
    protected final boolean _createDb;

    protected int _trace = 0;
    protected int _txTrace = 0;

    protected String _errorPrefix;

    protected boolean _deadlockWarning;

    protected DeactivateController _deactivateController = new DeactivateController();

    private Ice.Object _pingObject = new PingObject();
}
